src/SignalR/clients/csharp/Client.Core/src/Internal/InvocationRequest.cs

```diff
@@ -13,15 +15,17 @@ namespace Microsoft.AspNetCore.SignalR.Client.Internal;
 internal abstract partial class InvocationRequest : IDisposable
 {
     private readonly CancellationTokenRegistration _cancellationTokenRegistration;
+    private int _isActivityStopping;
```

Comment:
Did you find an issue with requests being completed/canceled multiple times?

Maybe if it was canceled by the user and completed by a response at the same time?

Reply:
`Cancel()` is always called from `Dispose()` regardless of whether `Complete()` or `Fail()` is already called. So that needed to be guarded against.

I wasn't sure whether these methods could happen simultaneously. All the other APIs called from them looked thread safe so I guessed probably and used Interlocked to make updating Activity thread safe.

A race between cancellation and complete/fail is the main scenario I thought of. I know that is possible in the gRPC client.

Reply:
> `Cancel()` is always called from `Dispose()` regardless of whether `Complete()` or `Fail()` is already called. So that needed to be guarded against.

By the way, this allocates an extra exception for streaming. `OperationCanceledException` is always allocated on invocation dispose, even if the stream was successfully completed.

https://github.com/dotnet/aspnetcore/blob/42e61729ecee15701ba55f4dbc6dc62d966a2a7f/src/SignalR/clients/csharp/Client.Core/src/Internal/InvocationRequest.cs#L124